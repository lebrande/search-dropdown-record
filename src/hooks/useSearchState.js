import { useState, useEffect, createContext } from 'react';
import { stateService } from '../services/state';
import { searchStates, getState } from '../utils/state';

export const SearchStateContext = createContext(null);

export const useSearchState = () => {
  const [states, setStates] = useState([]);
  const [query, setQuery] = useState('');
  const [resultsList, setResultsList] = useState([]);
  const [pickedStateCode, setPickedStateCode] = useState(null);
  const [searchKey, setSearchKey] = useState('state');
  const [dataType, setDataType] = useState('table');
  const currentState = getState(pickedStateCode, states);

  useEffect(() => {
    stateService.fetchAllStates().then(({ data }) => {
      setStates(data);
    })
  }, []);

  useEffect(() => {
    setResultsList(searchStates(query, searchKey, states));

    if (currentState && currentState.state !== query) {
      setPickedStateCode(null);
    }
  }, [query]);

  useEffect(() => {
    if (currentState) {
      setQuery(currentState.state);
    }
  }, [pickedStateCode]);

  return {
    query,
    onSetQuery: setQuery,
    list: resultsList,
    onPick: ({ code }) => setPickedStateCode(code),
    currentState,
    dataType,
    setSearchKey,
    setDataType,
  };
}