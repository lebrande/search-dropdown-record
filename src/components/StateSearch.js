import React, { useState, useEffect, useRef, useContext } from 'react';
import classnames from 'classnames';
import { useOnClickOrFocusOutside } from '../hooks/useOnClickOrFocusOutside';
import { SearchStateContext } from '../hooks/useSearchState';
import { getSelectedItemOnKey } from '../utils/getSelectedItemOnKey';

const StateSearch = () => {
  const {
    query,
    onSetQuery,
    list,
    onPick,
  } = useContext(SearchStateContext);
  const [selected, setSelected] = useState(0);
  const [focused, setFocused] = useState(false);
  const wrapperRef = useRef(null);

  const _onPick = (item) => {
    onPick(item);
    setFocused(false);
  }

  const pickOnKey = (key) => {
    if (key === 'Enter') {
      const selectedItem = list[selected];
      if (selectedItem) {
        _onPick(selectedItem);
      }
    }
  };

  useOnClickOrFocusOutside(
    wrapperRef,
    () => setFocused(false),
  );

  useEffect(() => {
    setSelected(0);
  }, [query]);

  return (
    <div className="field" ref={wrapperRef}>
      <div className="control">
        <div className="dropdown is-active">
          <div className="dropdown-trigger">
            <input
              data-testid="input"
              value={query}
              onChange={({ target }) => onSetQuery(target.value)}
              className="input"
              type="text"
              placeholder="wyszukaj..."
              onKeyUp={({ key }) => {
                setSelected(getSelectedItemOnKey(key, selected, list.length));
                pickOnKey(key);
              }}
              onFocus={() => setFocused(true)}
            />
          </div>
          {focused && list.length > 0 && (
            <div data-testid="dropdown-menu" className="dropdown-menu">
              <div className="dropdown-content">
                {list.map((item, index) => {
                  const { state } = item;

                  return (
                    <a
                      key={state}
                      data-testid="dropdown-item"
                      className={classnames('dropdown-item', {
                        'is-active': selected === index,
                      })}
                      onMouseEnter={() => setSelected(index)}
                      onClick={() => _onPick(item)}
                    >
                      {state}
                    </a>
                  );
                })}
              </div>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default StateSearch;
