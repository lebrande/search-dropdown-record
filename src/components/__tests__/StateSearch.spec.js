import React from 'react';
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer';
import StateSearch from '../StateSearch';
import { SearchStateContext } from '../../hooks/useSearchState';

describe('StateSearch', () => {
  it('default render matches snapshot', () => {
    const tree = renderer
      .create((
        <SearchStateContext.Provider value={{
          query: "",
          onSetQuery: jest.fn(),
          list: [],
          onPick: jest.fn(),
        }}>
          <StateSearch />
        </SearchStateContext.Provider>
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('dropdown list - show on click input and hide on select', () => {
    const testList = [
      { state: 'state1' },
      { state: 'state2' },
      { state: 'state3' },
    ];

    render((
      <SearchStateContext.Provider value={{
        query: "",
        onSetQuery: jest.fn(),
        list: testList,
        onPick: jest.fn(),
      }}>
        <StateSearch />
      </SearchStateContext.Provider>
    ))

    // check there is no dropdown list rendered
    const menuBefore = screen.queryAllByTestId('dropdown-menu');
    expect(menuBefore.length).toBe(0);

    // click input
    const input = screen.queryByTestId('input');
    userEvent.click(input);

    // dropdown appears
    const menuAfterClickInput = screen.queryAllByTestId('dropdown-menu');
    expect(menuAfterClickInput.length).toBe(1);

    // check items count
    const items = screen.queryAllByTestId('dropdown-item');
    expect(items.length).toBe(testList.length);

    // check state name is displayed
    expect(items[0].textContent).toBe(testList[0].state);

    // menu disappears when click (select) an item
    userEvent.click(items[0]);
    const menuAfterClickItem = screen.queryAllByTestId('dropdown-menu');
    expect(menuAfterClickItem.length).toBe(0);
  });
});
