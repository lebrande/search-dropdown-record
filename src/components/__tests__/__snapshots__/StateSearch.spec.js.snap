// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateSearch default render matches snapshot 1`] = `
<div
  className="field"
>
  <div
    className="control"
  >
    <div
      className="dropdown is-active"
    >
      <div
        className="dropdown-trigger"
      >
        <input
          className="input"
          data-testid="input"
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          placeholder="wyszukaj..."
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</div>
`;
